sql: arrays in pg_type get nonzero typelem columns

jordanlewis opened the pull request:
Arrays now have the `typelem` column of their entry in `pg_type` set to the oid of their element type. For instance, the `_int8` type (array of `int8`) has its `typelem` set to `int8`.

Resolves #13524.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13535)
<!-- Reviewable:end -->

jordanlewis committed:
sql: arrays in pg_type get nonzero typelem columns

Arrays now have the `typelem` column of their entry in `pg_type` set to
the oid of their element type. For instance, the `T__int8` type (array
of `int8`) has its `typelem` set to `T_int8`.

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-Kce2ls64TFi4yUDqG7G:bxevcpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
> ```Go
> }
> 
> // arrayOidToElemOid maps the OIDs of Postgres array types to the OIDs of the
> ```

I feel like we can be a bit smarter about this. If we exported `parser.tArray` then we could get rid of this mapping and write `typElem = typ.(parser.TArray).Typ.Oid()`.

This would require a slight change in `parser/type.go`. Instead of our `OidToType` map, we could take advantage of `tOidWrapper`. We could replace the entire map with an array that contains all types, including variants. The `AllTypes` array would look something like:

```
var (
	// .. normal exported types.
)

var (
	// unexported variant types
	typeInt2 = wrapTypeWithOid(TypeInt, oid.T_int2),
	typeInt2Array = tArray{typeInt2}
)

var AllTypes = [...]Type{
	TypeBool,
	TypeInt,
	typeInt2,
	TypeIntArray,
	typeInt2Array,
}
```

Using this approach, we could then also kill `aliasedOidToName` by extending `customOidNames` (though I think `aliasedOidToName` is a better name)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR!

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdCvF6eeA51ytskpr1O:ble7r6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I feel like we can be a bit smarter about this. If we exported `parser.tArray` then we could get rid of this mapping and write `typElem = typ.(parser.TArray).Typ.Oid()`.

This would require a slight change in `parser/type.go`. Instead of our `OidToType` map, we could take advantage of `tOidWrapper`. We could replace the entire map with an array that contains all types, including variants. The `AllTypes` array would look something like:

```
var (
	// .. normal exported types.
)

var (
	// unexported variant types
	typeInt2 = wrapTypeWithOid(TypeInt, oid.T_int2),
	typeInt2Array = tArray{typeInt2}
)

var AllTypes = [...]Type{
	TypeBool,
	TypeInt,
	typeInt2,
	TypeIntArray,
	typeInt2Array,
}
```

Using this approach, we could then also kill `aliasedOidToName` by extending `customOidNames` (though I think `aliasedOidToName` is a better name)
</blockquote></details>

I think we still need the OidToType map for when pgwire parses oids, and we still need `aliasedOidToName` because we report even the "base" types incorrectly sometimes, like `int[]` should be `int8[]` for `pg_catalog`.

I updated this to export `TArray` though and added the alias types as you suggested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-:-KdD4YIR0_NWug5Zlxe_:b-814g8k)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 6 of 10 files at r2.
Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdD8tazydIdGdrTKmJ6:b-jbk6tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think we still need the OidToType map for when pgwire parses oids, and we still need `aliasedOidToName` because we report even the "base" types incorrectly sometimes, like `int[]` should be `int8[]` for `pg_catalog`.

I updated this to export `TArray` though and added the alias types as you suggested.
</blockquote></details>

Good point about the `OidToType`.

I'm not sure I'm clear on the `aliasedOidToName` part, though. Won't the recursion in `TArray.SQLName()`, in conjunction with `tOidWrapper's` use of `customOidNames`, assure that we report the correct name? I'm just realizing the we use `Type.String` instead of `Type.SQLName` in this method. Maybe that's the issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdDAl8ogHbekpK05f1e:b65vi31) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Good point about the `OidToType`.

I'm not sure I'm clear on the `aliasedOidToName` part, though. Won't the recursion in `TArray.SQLName()`, in conjunction with `tOidWrapper's` use of `customOidNames`, assure that we report the correct name? I'm just realizing the we use `Type.String` instead of `Type.SQLName` in this method. Maybe that's the issue.
</blockquote></details>

Unfortunately, `Type.SQLName` is yet another kind of name that types need to support - the "sql standard" name. They are names like `bigint` and `character varying`.

I suppose we could introduce another method on Type. Thoughts? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdDUrACYMHvURhxON_2:b-ehqpd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Unfortunately, `Type.SQLName` is yet another kind of name that types need to support - the "sql standard" name. They are names like `bigint` and `character varying`.

I suppose we could introduce another method on Type. Thoughts? 
</blockquote></details>

Got it. Is there any reason not to have `Type.String` itself return the corresponding OID's name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdDb_ci_jkRQPrgqLAE:b-k1e2jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Got it. Is there any reason not to have `Type.String` itself return the corresponding OID's name?
</blockquote></details>

Wouldn't that introduce an inconsistency between how types are spelled in the language (`INT`) and how they'd be printed out in various places (`int8`)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdUaEEz57aJ_UZdwE3G:bwkph74) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Wouldn't that introduce an inconsistency between how types are spelled in the language (`INT`) and how they'd be printed out in various places (`int8`)?
</blockquote></details>

Yeah that's true. I'm fine either way then. A new method wouldn't be terrible, but it might be best just to leave it as you have it here for now. Up to you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-:-KdUagzQx6TLg9Q5dwQ4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 6 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535#-Kce2ls64TFi4yUDqG7F:-KdWn2jv1Rd-TGaloo8W:bikcj0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/68a746f08c2a15d8f87263f964b0a380d0916644/pkg/sql/pg_catalog.go#L1500)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah that's true. I'm fine either way then. A new method wouldn't be terrible, but it might be best just to leave it as you have it here for now. Up to you.
</blockquote></details>

I'm going to leave this as-is. I think there's probably a nice way to unify all of the different type names, but I don't see what it is right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13535)*
<!-- Sent from Reviewable.io -->
